File: 1326-minimum-number-of-taps-to-open-to-water-a-garden/1326-minimum-number-of-taps-to-open-to-water-a-garden.cpp
class Solution {
public:
    int minTaps(int n, vector<int>& ranges) {
        // Create a vector to store the minimum number of taps needed for each position
        vector<int> dp(n+1, 1e9);
        // Initialize the starting position of the garden
        dp[0] = 0;
        for(int i=0; i<=n; i++){
            // Calculate the leftmost position reachable by the current tap
            int tapStart = max(0, i-ranges[i]);
            // Calculate the rightmost position reachable by the current tap
            int tapEnd = min(n, i+ranges[i]);
            for(int j=tapStart; j<=tapEnd; j++){
                // Update with the minimum number of taps
                dp[tapEnd] = min(dp[tapEnd], dp[j]+1);
            }
        }
        if(dp[n] == 1e9) return -1;
        return dp[n];
    }
};

File: 1326-minimum-number-of-taps-to-open-to-water-a-garden/1326-minimum-number-of-taps-to-open-to-water-a-garden.py
class Solution:
    def minTaps(self, n: int, ranges: List[int]) -> int:
        maxReach = [0] * (n+1)
        for i in range(0, len(ranges)):
            start = max(0, i-ranges[i])
            end = min(n, i+ranges[i])
            maxReach[start] = max(maxReach[start], end)
        taps, currEnd, nextEnd =0, 0, 0
        for i in range(0, n+1):
            if i > nextEnd:
                return -1
            if i > currEnd:
                taps+=1
                currEnd = nextEnd
            nextEnd = max(nextEnd, maxReach[i])
        return taps